proposals/tf-operator-design-v1alpha2.md

```diff
@@ -0,0 +1,435 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*
+
+- [TF-Operator Design (v1alpha2)](#tf-operator-design-v1alpha2)
+  - [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [UI or API](#ui-or-api)
+  - [Design](#design)
+    - [TFController](#tfcontroller)
+    - [Distributed TensorFlow Configuration](#distributed-tensorflow-configuration)
+    - [Event-Driven](#event-driven)
+    - [Reconciler](#reconciler)
+    - [Error Handling](#error-handling)
+    - [Test](#test)
+  - [Alternatives Considered](#alternatives-considered)
+    - [Future Works](#future-works)
+    - [Related Issues](#related-issues)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+_Authors:_
+
+* @ScorpioCPH - Penghao Cen &lt;cenph@caicloud.io&gt;
+
+# TF-Operator Design (v1alpha2)
+
+## Motivation
+
+Kubeflow community currently have [tf-operator](https://github.com/kubeflow/tf-operator) (v1alpha1) for running TensorFlow jobs on Kubernetes. And we have received some refactoring [requests](https://github.com/kubeflow/tf-operator/issues?q=is%3Aissue+is%3Aopen+label%3Akind%2Fapi-change) for API changes.
+Open this file to summarize the design details and move the version of API to `v1alpha2`.
+
+## Goals
+
+- Define the structure of API `v1alpha2`.
+    + Cover most of the refactoring requests we have discussed.
+    + Simplify the API definition.
+- Define an `event-driven` mechanism for TFJob life-cycle management.
```

Comment:
Would it make sense to split the implementation into its own proposal as opposed to using 1 proposal for the implementation and API?

Reply:
I think they are bundled. If we do not migrate the implementation to the new design we have to implement twice, once for etcd-operator style and once for event-driven style. I think it is unnecessary. :thinking:

Reply:
Hi, this proposal follows the [proposal-template](https://github.com/kubeflow/community/blob/master/proposals/TEMPLATE.md), it makes sense for me to put API and design together in one proposal :)

Reply:
@gaocegege @ScorpioCPH  Splitting the proposals doesn't mean implementing them twice. The motivation for splitting the proposals is to narrow the scope of each proposal which should speedup agreement. Actual work doesn't need to begin until both proposals are approved.

I'd expect the API to largely be independent of the implementation. So if we move implementation into its own proposal, that will focus discussion on the API and help us converge.

Reply:
@jlewi Hi, i don't think it a good idea to split one design doc into two pieces.
Keep this in one doc is better for reader to find the details in on place.

Reply:
Sure.